File: AlgorithmicToolbox/week4_divide_and_conquer/4_improving_quicksort/4_improving_quicksort.cpp
#include <iostream>
#include <vector>
#include <cstdlib>

using std::vector;
using std::swap;

/*
The goal in this problem is to redesign a given implementation of the randomized quick sort
algorithm so that it works fast even on sequences containing many equal elements.
*/

void partition3(vector<int>& a, int l, int r, int &p1, int &p2) {
    int x = a[l], i = l;
    p1 = l;
    p2 = r;
    
    while (i <= p2) {
        if (a[i] < x) {
            swap(a[i], a[p1]);
            p1++;
            i++;
        }
        else if (a[i] > x) {
            swap(a[i], a[p2]);
            p2--;
        }
        else {
            i++;
        }
    }
}

void randomized_quick_sort(vector<int>& a, int l, int r) {
    if (l >= r) {
        return;
    }

    int k = l + rand() % (r - l + 1);
    swap(a[l], a[k]);
    int m1, m2;
    partition3(a, l, r, m1, m2);

    randomized_quick_sort(a, l, m1 - 1);
    randomized_quick_sort(a, m2 + 1, r);
}

int main() {
    int n;
    std::cin >> n;
    vector<int> a(n);
    for (size_t i = 0; i < a.size(); ++i) {
        std::cin >> a[i];
    }
    randomized_quick_sort(a, 0, a.size() - 1);
    for (size_t i = 0; i < a.size(); ++i) {
        std::cout << a[i] << ' ';
    }
}

File: AlgorithmicToolbox/week4_divide_and_conquer/4_improving_quicksort/4_improving_quicksort.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


"""
The goal in this problem is to redesign a given implementation of the randomized quick sort
algorithm so that it works fast even on sequences containing many equal elements.
"""
# This 3-way partitioning uses the Dutch National Flag problem algorithm.

from random import randint

def partition3(A, l, r):
    
    lt = l  
    i = l   
    gt = r  
    pivot = A[l]   
    while i <= gt:
        
        if A[i] < pivot:
            A[lt], A[i] = A[i], A[lt]
            lt += 1
            i += 1
        elif A[i] > pivot:
            A[i], A[gt] = A[gt], A[i]
            gt -= 1
        else:
            i += 1
            
    return lt, gt


def randomized_quick_sort(array, l, r):
    if l >= r:
        return array
    
    k = randint(l, r)
    array[l], array[k] = array[k], array[l]
    
    m1, m2 = partition3(array, l, r)
    
    randomized_quick_sort(array, l, m1-1)
    randomized_quick_sort(array, m2+1, r)


if __name__ == '__main__':
    input_n = int(input())
    elements = list(map(int, input().split()))
    assert len(elements) == input_n
    randomized_quick_sort(elements, 0, len(elements) - 1)
    print(*elements)


# In[ ]:




